# Finding brain regions and getting their center of mass using the Allen SDK

1. Create a new python virtual environment or use an existing one.
1. pip install allensdk==2.15.1 # this works with the current ubuntu/pipeline we have, you can try just doing pip install allensdk
1. The Allen data has to be download which happens in the commands below. You might get an error on downloading, if you do, just copy the link that is in the error message and use curl or wget or just put that link in the browser. It will then download. You will need to move that into the current working directory of this notebook under the 'mouse_connectivity' directory
1. You'll notice the atlas below has the rostrum at the top. These are the coordinates of the Allen atlas. We need the rostrum at the left, with the superior at the top in the sagittal view. So the center of mass coordinates you get below need to be flipped.


In [2]:
import numpy as np
import pandas as pd
from scipy.ndimage import center_of_mass
from allensdk.core.mouse_connectivity_cache import MouseConnectivityCache
from matplotlib import pyplot as plt

ModuleNotFoundError: No module named 'allensdk'

In [3]:
! pip install allensdk

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.0/4.0 MB 590.8 kB/s eta 0:00:00m eta 0:00:010:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 1.2 MB/s eta 0:00:00m eta 0:00:010:01:01
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.6/4.6 MB 875.3 kB/s eta 0:00:00m eta 0:00:010:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 54.5/54.5 kB 1.4 MB/s eta 0:00:001.6 MB/s eta 0:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.8/17.8 MB 2.3 MB/s eta 0:00:00m eta 0:00:010:00:01
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 37.3/37.3 MB 2.3 MB/s eta 0:00:00m eta 0:00:010:00:01
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 131.9/131.9 kB 2.5 MB/s eta 0:00:00 MB/s eta 0

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.1/49.1 kB 2.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 109.5/109.5 kB 2.3 MB/s eta 0:00:00 MB/s eta 0:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 131.7/131.7 kB 2.1 MB/s eta 0:00:002.4 MB/s eta 0:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 131.6/131.6 kB 2.0 MB/s eta 0:00:00 MB/s eta 0:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 126.6/126.6 kB 1.9 MB/s eta 0:00:002.8 MB/s eta 0:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 233.8/233.8 kB 2.5 MB/s eta 0:00:003.1 MB/s eta 0:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 485.6/485.6 kB 2.7 MB/s eta 0:00:00 MB/s eta 0:00:01:01
  Created wheel for aiohttp: filename=aiohttp-3.7.4-cp310-cp310-linux_x86_64.whl size=1273573 sha256=0e1843fcafaa69c3806172623539f68e5dbbc3aaad7f1865c90bcb30e78bb0f5
  Stored in directory: /home/eddyod/.cache/pip/wheels/a0/67/11/5beda0150ba642c0c5475964d4c5e00b8a2b50a19e3aaab8ac
  Created wheel for glymur: filename=Gly

  Attempting uninstall: botocore
    Found existing installation: botocore 1.27.64
    Uninstalling botocore-1.27.64:
      Successfully uninstalled botocore-1.27.64
  Attempting uninstall: aiohttp
    Found existing installation: aiohttp 3.8.1
    Uninstalling aiohttp-3.8.1:
      Successfully uninstalled aiohttp-3.8.1
  Attempting uninstall: s3transfer
    Found existing installation: s3transfer 0.6.0
    Uninstalling s3transfer-0.6.0:
      Successfully uninstalled s3transfer-0.6.0
  Attempting uninstall: boto3
    Found existing installation: boto3 1.24.64
    Uninstalling boto3-1.24.64:
      Successfully uninstalled boto3-1.24.64


In [79]:
%matplotlib notebook

### Accessing the CCF from the MouseConnectivityCache

Instantiate the MouseConnectivityCache (mcc). The default resolution is 25 microns, but 10,50, and 100 microns are also valid options. You can also specify a location of a manifest file, which would keep track of all downloaded assets from template and annotation volumes to experimental data. If a manifest file directory is not provided, one will be created in the current working directory. (This is where you might get downloading errors, just download it manually and place it under the dir: mouse_connectivity'

In [80]:
mcc = MouseConnectivityCache(resolution=25)

#### Annotated Volume and Reference Space
Now you can get the Reference Space (rsp) at that resolution. This automatically downloads the annotated volume at the specified resolution (25 microns).

In [81]:
rsp = mcc.get_reference_space()

The annotated volume is a 3D numpy ndarray, with axes correspond to AP,DV, and ML respectively:

In [82]:
print('Type of annotated volume', type(rsp))
print('Shape of annotated volume:',rsp.annotation.shape)
print('Annotated volume data type:',rsp.annotation.dtype)

Type of annotated volume <class 'allensdk.core.reference_space.ReferenceSpace'>
Shape of annotated volume: (528, 320, 456)
Annotated volume data type: uint32


The rsp provides access to the structure tree from the ontology. Given that there are multiple atlases and associated ontologies, we are only interested in the structures belonging to this annotation. Thus, we remove unassigned structures from the structure tree. This returns a structure graph which is more easily viewed as a pandas DataFrame:

In [83]:
sg = rsp.remove_unassigned()

In [84]:
sg = pd.DataFrame(sg)

In [85]:
sg.head()

acronym  graph_id  graph_order   id                           name  \
0    root         1            0  997                           root   
1    grey         1            1    8  Basic cell groups and regions   
2      CH         1            2  567                       Cerebrum   
3     CTX         1            3  688                Cerebral cortex   
4   CTXpl         1            4  695                 Cortical plate   

         structure_id_path                                  structure_set_ids  \
0                    [997]                                        [691663206]   
1                 [997, 8]  [112905828, 691663206, 12, 184527634, 11290581...   
2            [997, 8, 567]  [112905828, 691663206, 12, 184527634, 11290581...   
3       [997, 8, 567, 688]  [112905828, 691663206, 12, 184527634, 11290581...   
4  [997, 8, 567, 688, 695]  [112905828, 691663206, 12, 184527634, 11290581...   

       rgb_triplet  
0  [255, 255, 255]  
1  [191, 218, 227]  
2  [176, 240, 255]  
3  [176, 255, 184]  
4  [112, 255, 112]

In [92]:
# I picked the cerebellum as it is easy to see
# get data with structure name
structure = 'SC'
row = sg.loc[sg['name'] == structure]
row

Empty DataFrame
Columns: [acronym, graph_id, graph_order, id, name, structure_id_path, structure_set_ids, rgb_triplet]
Index: []

In [86]:
# get data with acronym
acronym = 'SCs'
row = sg.loc[sg['acronym'] == acronym]
row

acronym  graph_id  graph_order   id                                  name  \
536     SCs         1          808  302  Superior colliculus, sensory related   

                structure_id_path  \
536  [997, 8, 343, 313, 339, 302]   

                                     structure_set_ids      rgb_triplet  
536  [112905828, 691663206, 687527945, 10, 12, 6881...  [255, 122, 255]

In [1]:
sg.query('name.str.contains("facial")', engine='python')

NameError: name 'sg' is not defined

In [87]:
structure_id = row.iat[0, 3]

In [88]:
# Get a list of acronymns
id_acronym_map = rsp.structure_tree.get_id_acronym_map() # Dictionary returns IDs given acronyms
acronym_id_map = {v:k for k,v in id_acronym_map.items()} # Flip key:value pairs to get dictionary for acronyms given IDs
print(list(map(acronym_id_map.get,rsp.structure_tree.descendant_ids([structure_id])[0])))

['SCs', 'SCop', 'SCsg', 'SCzo']


The structures' acronyms, names and IDs can all be linked through the structure graph.

While the rsp can be used to generate sections in the coronal (0), horizontal (1), or sagittal (2) planes:

In [89]:
orientation = {'coronal':0,'horizontal':1,'sagittal':2} # makes orientation arguments more readible
pos_microns = lambda x:x*25 # get_slice_image function takes positional arguments in microns instead of 25 micron increments

img = rsp.get_slice_image(orientation['sagittal'],pos_microns(200))
plt.imshow(img)

The rsp can also generate masks of given structure IDs. The IDs are provided as a list, and by default include all voxels annotated as children of the given structures (direct_only=False). If only the parent level voxels are desired, set direct_only to True.

Here we make a mask of the Cerebral Cortex, which from the structure graph is shown to have the ID 688.

In [90]:
structure_mask = rsp.make_structure_mask([structure_id], direct_only=False)

In [93]:
print('Shape of entire brain', rsp.annotation.shape)
print('Shape of mask:',structure_mask.shape)
print('Mask data type:',structure_mask.dtype)
values, counts = np.unique(structure_mask, return_counts=True)
com = center_of_mass(structure_mask)
print(f'Center of mass for {structure}: {com}')

Shape of entire brain (528, 320, 456)
Shape of mask: (528, 320, 456)
Mask data type: uint8
Center of mass for SC: (368.08973894913635, 62.21607279887645, 227.68008600344592)


In [77]:
com_section = int(round(com[2])) 
plt.scatter([com[1]], [com[0]], c='r', s=40)
plt.imshow(structure_mask[:,:,com_section])